pkg/util/traceevent/traceevent.go

```diff
@@ -0,0 +1,509 @@
+	value := eventSink.Load()
+	if value == nil {
+		return nil
+	}
+	return value.(sinkHolder).sink
+}
+
+// FlightRecorder returns the always-on in-memory recorder.
+func FlightRecorder() *RingBufferSink {
+	return flightRecorder
+}
+
+// TraceEvent records a trace event if the category is enabled.
+// The caller is responsible for applying any necessary redaction to the supplied fields.
+//
+// Usage:
+//
+//	traceevent.TraceEvent(traceevent.CoprRegionCache, ctx, "region split detected",
+//		zap.Uint64("regionID", regionID),
+//		zap.String("key", formatKey(key)))
+func TraceEvent(ctx context.Context, category TraceCategory, name string, fields ...zap.Field) {
+	if !IsEnabled(category) {
+		return
+	}
+
+	// Defensive copy prevents corruption from caller's buffer reuse.
+	// Reserve extra capacity for LogSink to append metadata without allocation.
+	event := Event{
+		Category:  category,
+		Name:      name,
+		Timestamp: time.Now(),
+		TraceID:   extractTraceID(ctx),
+		Fields:    copyFieldsWithCapacity(fields, 3),
+	}
+
+	// Record to flight recorder if enabled (base or full mode).
+	if recorderEnabled.Load() {
+		if recorder := FlightRecorder(); recorder != nil {
+			recorder.Record(ctx, event)
+		}
+	}
+
+	// Record to log sink if logging is enabled (full mode).
+	if sink := CurrentSink(); sink != nil {
+		sink.Record(ctx, event)
+	}
+}
+
+// extractTraceID returns the trace identifier from ctx if present.
+// It delegates to client-go's TraceIDFromContext implementation.
+func extractTraceID(ctx context.Context) []byte {
+	return trace.TraceIDFromContext(ctx)
+}
+
+// GenerateTraceID creates a trace ID from transaction start timestamp and statement count.
+// The trace ID is 20 bytes: [start_ts (8 bytes)][stmt_count (8 bytes)][random (4 bytes)] in big-endian format.
+// The random suffix distinguishes different statement executions.
+// This function should be called ONCE per statement execution, not per retry.
+// If no transaction has started, start_ts will be 0.
+func GenerateTraceID(startTS uint64, stmtCount uint64) []byte {
+	traceID := make([]byte, 20)
+	binary.BigEndian.PutUint64(traceID[0:8], startTS)
+	binary.BigEndian.PutUint64(traceID[8:16], stmtCount)
+	binary.BigEndian.PutUint32(traceID[16:20], rand.Uint32())
+	return traceID
+}
+
+// LogSink serializes trace events to the global zap logger. The output structure
+// stays simple for now, but the Event data contains enough information to build
+// a Trace Event Format record when the format is finalized.
+type LogSink struct{}
+
+// Record implements the Sink interface.
+func (*LogSink) Record(ctx context.Context, event Event) {
+	if !loggingEnabled.Load() {
+		return
+	}
+
+	// Append to reserved capacity without allocation.
+	// Field order: [event fields] [category] [timestamp] [trace_id?]
+	fields := event.Fields
+	fields = append(fields, zap.String("category", getCategoryName(event.Category)))
+	fields = append(fields, zap.Time("event_ts", event.Timestamp))
```

Comment:
How about display it as a number in microsecond? https://pkg.go.dev/time#Time.UnixMicro
zap.Time will be displayed as millisecond
We already has the time information in millisecond in log, but millisecond is too corase-grained